### ISSUE TITLE
Question about optimizer in optimize.py

### ISSUE THREAD

**Issue by dayanbatuofu** [Tenavi/HJB_NN#4]
Hi Tenavi,

Thanks for sharing your code and give me a nice answer to my previous confusion. Now I have one question about your code：  

I notice you have `def _minimize` in the `optimize.py`, but I could not find any place you use in the code. So I am curious about how you use this function. 

Thanks,

Lei Zhang

**Comment by Tenavi** [comment]
Hi Lei,

Of course, I'm happy to answer any questions. Please also feel free to email me directly at tenakamu@ucsc.edu.

The optimize.py file is based of some code from [TensorFlow `ExternalOptimizerInterface`](https://github.com/tensorflow/tensorflow/blob/r1.11/tensorflow/contrib/opt/python/training/external_optimizer.py) module. I'll make sure to update the commenting in that file to reflect this.

`_minimize` is called by `minimize` from the superclass `ExternalOptimizerInterface`, and my wrapping of the interface gets used in `_train_L_BFGS_B` in `utilities/neural_networks.py` (see line 347 in that file).

Does that answer your question?

Thanks,
Tenavi

**Comment by dayanbatuofu** [comment]
Hi Tenavi,

Thanks for your quick reply. I noticed the `_train_L_BFGS_B` before sending the question to you. There exists `optimizer.minimize`. 
But it looks different from the `_minimize`. So that is why I am so confused about how you use `_minimize`. I think when using `optimizer.minimize`, the code will also run `_minimize`. Please correct me if my understanding is wrong. BTW, I want to use pytorch to implement a similar method you wrote in the paper. But for the optimizer `L-BFGS-B`, I have no idea here because you modified some structure of class `ScipyOptimizerInterface` from superclass `ExternalOptimizerInterface`. Moreover, there exist some issues for the `L-BFGS-B` buildin function in the Pytorch after I search online. Maybe SGD should be an appropriate choice because you mentioned this method in the paper. So do you have any suggestions?

Thanks,

Lei Zhang

**Comment by Tenavi** [comment]
If you're doing this in PyTorch I wouldn't worry about the way I'm doing it. I had to do some modifications of the TensorFlow code so that the L-BFGS interface could accept pre-computed gradient graphs. Since PyTorch is all computed dynamically anyways I don't think what I did is necessary.

Regarding SGD (and things like Adam), I've found it to perform rather poorly compared to L-BFGS on this problem. Since we are dealing with relatively small NNs and data sets, the computational cost of L-BFGS is offset by the much higher convergence rate. Of course, this is only my experience and depending on what you do with it, you may find other optimizers do better.

**Comment by dayanbatuofu** [comment]
Thanks for your advice. It is very helpful.
